### ISSUE TITLE
show spinner in python function is deferred?

### ISSUE THREAD

**Issue by ay-ex** [pyscript/pyscript#2135]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [ ] I already searched in Google and didn't find any good information or help

### What happened?

thank you so much for the awesome pyscript project! 🙂 

on pyscript_2024.7.1(pyodide) (on brave browser, chromium engine), I encountered a strange behaviour(/bug?):

on a very simple page with just a spinner and two buttons show/hide spinner and a file load input, 
this happens (code attached below):

turning on/off the spinner via the buttons works exactly as expected. 
in function `get_file` turning on the spinner, then load a file, then turn of the spinner also seems to work as expected.

but as soon as I want to involve the spinner in a function, which processes the file 
(here: ifc file is read and elements inside are categorized into a dict) 
turning on the spinner at the top of the function seems to has no effect until the end of function. 
instead first the file is processed, ifc results are printed to console and then after(!) that, the spinner call is run.  🤔 

which in this case is obviously not so helpful, as the intention is to indicate with the spinner to the user, 
that the slow process of processing the ifc file is happening and disable it afterwards.

it almost seems, as if the dom modification for the spinner was deferred to the end of the function scope (?) 
is this observed correctly? and is this the expected behaviour?
with pyscript how would I be able to show spinner first, then do the file processing, and afterwards disable the spinner?
(ideally I could just put decorator on slow functions with the id of the spinner to be shown/hidden)

thank you in advance.

p.s.: 
"I already searched in Google and didn't find any good information or help *" - 
^^ I used user-respecting search engines instead. 
sad that I have to tick this  to be able to submit this issue. 
maybe it  could be framed more neutral as web search instead?

### What browsers are you seeing the problem on? (if applicable)

Other

### Console info

```shell
console output looks in order as expected - the show_spinner is called first:  
`show_spinner: startup_spinner.hidden=True`
`startup_spinner.hidden=False`
...
`ifc scan complete`
and after that the ifc file scan is performed.
```

### Additional Context

```python
import asyncio
import os
import pathlib

import ifcopenshell
import js
import pyodide
from pyscript.web import dom
from pyscript.web.elements import br, button, details, div, h4, h5, input_, label, option
from pyscript.web.elements import p, select, summary, table, td, th, tr


def hide_spinner(event):
    print(f"hide_spinner: {startup_spinner.hidden=}")
    startup_spinner.hidden = True
    print(f"{startup_spinner.hidden=}")


def show_spinner(event):
    print(f"show_spinner: {startup_spinner.hidden=}")
    startup_spinner.hidden = False
    print(f"{startup_spinner.hidden=}")


async def get_file(event):
    print("loading ifc file:")
    show_spinner(None)
    files = event.target.files.to_py()
    for file in files:
        print(file.name)
        file_content = await file.text()
        if file.name.endswith(".ifc"):
            ifc_path = here_dir / file.name
            with open(ifc_path, 'w') as f:
                f.write(file_content)
                print(f"{ifc_path} written.")
                files_to_process["ifc"] = ifc_path
        else:
            print("ERROR: file is not an ifc file")
            files_to_process["ifc"] = None
    print("load file finished")
    hide_spinner(None)


async def scan_ifc_contents(event):
    print("scan_ifc_contents:")
    if not files_to_process.get("ifc"):
        print("not all files for file types: [icf] loaded")
        return
    show_spinner(None)
    print(f"scanning: {files_to_process['ifc']}")

    ifc_path  = files_to_process["ifc"]
    ifc_model = ifcopenshell.open(str(ifc_path))
    dom["#ifc_schema"][0].innerHTML = ifc_model.schema

    for element in ifc_model.by_type("IfcElement"):
        class_name = element.is_a()
        if class_name not in ifc_elements_by_ifc_class:
            ifc_elements_by_ifc_class[class_name] = []
        ifc_elements_by_ifc_class[class_name].append(element)

        psets = ifcopenshell.util.element.get_psets(element)
        if class_name not in props_by_pset_by_class:
            props_by_pset_by_class[class_name] = {}
        if class_name not in prop_count_by_class:
            prop_count_by_class[class_name] = 0
        for pset_name, pset in psets.items():
            if pset_name not in props_by_pset_by_class[class_name]:
                props_by_pset_by_class[class_name][pset_name] = set()
            for prop_name in pset:
                props_by_pset_by_class[class_name][pset_name].add(prop_name)
                prop_count_by_class[class_name] += 1

    for i, class_name in enumerate(sorted(ifc_elements_by_ifc_class), 1):
        instances = ifc_elements_by_ifc_class[class_name]
        print(class_name, len(instances))
        ifc_class_names_by_index[str(i)] = class_name

    for i, class_name in ifc_class_names_by_index.items():
        dom["#dd_ifc_classes"][0].append(option(class_name, value=i))

    selector = dom["#pset_selection"][0]

    for class_name, psets in props_by_pset_by_class.items():
        selector.append(h4(class_name))
        selector.append(h4("pset properties:"))
        for pset_name, pset in psets.items():
            selector.append(h5(pset_name))
            for prop_name in pset:
                check_box = input_()
                check_box.type = "checkbox"
                dom["#pset_selection"][0].append(check_box)
                dom["#pset_selection"][0].append(label(prop_name))
                dom["#pset_selection"][0].append(br())

    # hide_spinner(None)
    print("ifc scan complete")


ifc_class_dropdown = dom["#dd_ifc_classes"][0]
page_title = dom["a#page-title"]
header_title = dom["title#header-title"]
startup_spinner = dom["#startup_spinner"][0]
table_spinner = dom["#table_spinner"][0]

ifc_elements_by_ifc_class = {}
ifc_class_names_by_index = {}
props_by_pset_by_class = {}
prop_count_by_class = {}

here_dir = pathlib.Path(os.getcwd())
files_to_process = {"ifc": None}

get_file_proxy = pyodide.ffi.create_proxy(get_file)
js.document.getElementById("fileinput").addEventListener("change", get_file_proxy)


print(f"loaded: {ifcopenshell.version=}")

hide_spinner(None)
print("page loaded")
```

**Comment by WebReflection** [comment]
@ay-ex it's hard to help anyhow here because we miss your config and the layout and until we are unable to reproduce the whole issue there won't be progress, I am afraid.

**Comment by ay-ex** [comment]
@WebReflection thank you for your response. 
so generally you would say it should be possible to set a spinner to running, then perform some work with a file from the virtual files system with a python package and turn the spinner off after that task is finished?

the toml config looks like this: 
```toml
name = "ifcos_elem_table"
description = "A basic application testing ifcopenshell (ifcos) in pyscript."
packages = [
    "/static/py/IfcOpenShell-0.7.0-py3-none-any.whl",
]
```
the above mentioned ifcopenshell wheel is this one: 
https://github.com/IfcOpenShell/wasm-preview/blob/master/IfcOpenShell-0.7.0-py3-none-any.whl

the html looks like this:
```html
<!doctype html>
<html>
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width,initial-scale=1.0">
        <link rel="stylesheet" href="/static/js/pyscript_2024.7.1/core.css">
        <link rel="stylesheet" href="/static/css/site.css">

        <script type="module" src="/static/js/pyscript_2024.7.1/core.js"></script>
        <title id="header-title"></title>
        <link rel="shortcut icon" type="image/png" href="/static/img/favicon.png"/>
    </head>

    <body>

        <div class="app-header"><a class="title" id="page-title" href="" style="color: #f0ab3c"></a></div>
        <div id="startup_spinner" class="loader"></div>

    <button
        type="submit"
        id="btn-hide_spinner"
        class="py-button"
        py-click="hide_spinner">
        hide_spinner
    </button>

    <button
        type="submit"
        id="btn-show_spinner"
        class="py-button"
        py-click="show_spinner">
        show_spinner
    </button>


        <section id="section_file_loading">
            <h3>ifc file content:</h3>
            <label for="fileinput">select ifc file:</label>
            <input type="file" id="fileinput" name="fileInput"/>
            <br>
            <div id="file_spinner" class="loader" hidden></div>
        </section>

        <section id="section_ifc_scanning">
            <h3>scan ifc file content:</h3>
            <button
                type="submit"
                id="btn-scan-ifc"
                class="py-button"
                py-click="scan_ifc_contents">
                scan ifc contents
            </button>
            <div class="key_value">
              <p class="leftContent">ifc schema:</p><p class="rightContent" id="ifc_schema"></p>
            </div>
            <div id="ifc_spinner" class="loader" hidden></div>
        </section>


        <section id="section_pset_selection">
            <br>
                <details>
                <summary>Pset selection</summary>
                    <div id="pset_selection"></div>
                </details>
            <br>
        </section>


        <section id="section_elem_table">
            <h3 id="elem_table_h3">ifc element table:</h3>

            <div id="ifc_class_buttons">
                <select id="dd_ifc_classes">
                </select>
                <button
                    type="submit"
                    id="btn-list-ifc-elements"
                    class="py-button"
                    py-click="list_ifc_elements">
                    list_ifc_elements
                </button>
            </div>
            <br>

            <div id="ifc_elem_table"></div>
            <div id="table_spinner" class="loader" hidden></div>
        </section>


        <section class="pyscript">
            <script type="py"
                    src="/static/py/pyscript_ifcos_elem_table.py"
                    config="/static/py/pyscript_ifcos_elem_table.toml">
            </script>
        </section>

    </body>
</html>

```

**Comment by WebReflection** [comment]
> generally you would say it should be possible to set a spinner to running, then perform some work with a file from the virtual files system with a python package and turn the spinner off after that task is finished?

the issue usually is with blocking operations ... until your current callback blocks there won't be any chance to see anything on the DOM and this is true with just JavaScript too.

A good way is usually to show the spinner and then schedule some async task to then, once that's completed, hide the spinner again.

**Comment by WebReflection** [comment]
@ay-ex I also miss `/static/css/site.css` do I am not sure I am seeying what you are seeing

**Comment by WebReflection** [comment]
@ay-ex never mind, I wrote some text in there so no CSS needed ... this works to me:

```python

async def scan_ifc_contents_blocking(event):
    print(f"scanning: {files_to_process['ifc']}")

    ifc_path  = files_to_process["ifc"]
    ifc_model = ifcopenshell.open(str(ifc_path))
    dom["#ifc_schema"][0].innerHTML = ifc_model.schema

    for element in ifc_model.by_type("IfcElement"):
        class_name = element.is_a()
        if class_name not in ifc_elements_by_ifc_class:
            ifc_elements_by_ifc_class[class_name] = []
        ifc_elements_by_ifc_class[class_name].append(element)

        psets = ifcopenshell.util.element.get_psets(element)
        if class_name not in props_by_pset_by_class:
            props_by_pset_by_class[class_name] = {}
        if class_name not in prop_count_by_class:
            prop_count_by_class[class_name] = 0
        for pset_name, pset in psets.items():
            if pset_name not in props_by_pset_by_class[class_name]:
                props_by_pset_by_class[class_name][pset_name] = set()
            for prop_name in pset:
                props_by_pset_by_class[class_name][pset_name].add(prop_name)
                prop_count_by_class[class_name] += 1

    for i, class_name in enumerate(sorted(ifc_elements_by_ifc_class), 1):
        instances = ifc_elements_by_ifc_class[class_name]
        print(class_name, len(instances))
        ifc_class_names_by_index[str(i)] = class_name

    for i, class_name in ifc_class_names_by_index.items():
        dom["#dd_ifc_classes"][0].append(option(class_name, value=i))

    selector = dom["#pset_selection"][0]

    for class_name, psets in props_by_pset_by_class.items():
        selector.append(h4(class_name))
        selector.append(h4("pset properties:"))
        for pset_name, pset in psets.items():
            selector.append(h5(pset_name))
            for prop_name in pset:
                check_box = input_()
                check_box.type = "checkbox"
                dom["#pset_selection"][0].append(check_box)
                dom["#pset_selection"][0].append(label(prop_name))
                dom["#pset_selection"][0].append(br())

    # hide_spinner(None)
    print("ifc scan complete")

def scan_ifc_contents(event):
    print("scan_ifc_contents:")
    if not files_to_process.get("ifc"):
        print("not all files for file types: [icf] loaded")
        return
    show_spinner(None)

    from asyncio import create_task
    create_task(scan_ifc_contents_blocking(event)).add_done_callback(lambda _:hide_spinner(None))

```

you can see I've moved out the blocking/heavy part and you see the spinner instantly visible and then gone once that heavy function is done.

The reason we introduced workers though is to avoid blocking the main thread for cases like these ... it won't play well with inline `py-click` attributes because the worker exists somewhere else but from the worker you can setup listeners too so I suggest, in the near future, to use workers.

Also keep in mind `pyscript.web` `dom` is gone so your code will break with 2024.8.1 and I am going to cc @ntoll and @mchilvers here because I did see these kind of issues coming.

Anyway, I am going to close this as it's not PyScript fault you block the whole thread, the solution is hopefully easy to reason about or understand (schedule a task, don't block forever) and the ideal world is to use `worker` instead where none of this would be an issue because when a worker changes the DOM it happens inevitably in a worker blocking way but never in a main thread blocking way so you won't need to schedule tasks manually.

**Comment by ay-ex** [comment]
@WebReflection : amazing 😌 , thank you so much for your explanations, recommendations and snippets - works like a charm! 🪄 
agreed this is not a pyscript bug, but instead a blocking - non-blocking topic.
I will read up on worker and dom to see how to improve there.
